### ISSUE TITLE
ModuleFinder incorrectly assumes symlinked packages live in same-name folder.

### ISSUE THREAD

**Issue by dcporter** [Galooshi/import-js#379]
`ModuleFinder`’s helper method `assumedLocalPath` makes a bad assumption that symlinked packages live in folders that precisely match their packages. For example, it assumes that a symlinked "mobx" package lives in a folder called precisely `/mobx/`. It's trivial to break this with:

```sh
$ cd /Users/dcporter/Code
$ git clone https://github.com/mobxjs/mobx.git mobx-local
$ cd mobx-local
$ npm link
$ cd ../test-project
$ npm link mobx  # ./node_modules/mobx is now a symlink to a folder called "mobx-local".
```

In this case, `require.resolve` returns `/Users/dcporter/Code/mobx-local/lib/mobx.js`, and `assumedLocalPath` turns that into the nonsensical `"./node_modules//Users/dcporter/Code/mobx-local/lib/mobx.js"`, which super doesn't exist and which fails loudly farther down the line.

The "local checkout with a different name" problem is also likely to be triggered with [scoped packages](https://docs.npmjs.com/misc/scope), whose package names start with "@" and which are nested inside scope folders, e.g. `node_modules/@my-company/package/`; ~~and it's effectively unavoidable if you're using [Lerna](https://github.com/lerna/lerna) to manage a monorepo.~~

A solution is to (recursively) follow the symlink to get the actual target directory, and proceed from there. In the above example, `./node_modules/mobx` points eventually to `/Users/dcporter/Code/mobx-local/`, so that can be used to correctly extract the intra-package part of the path.

I threw together some test code locally and found the following useful:
- `npm link` sets up a two-tiered symlink — one from the linked package into `/usr/local/bin`, and one from there into the project which is using it — so you have to recursively follow the symlink.
- `fs.readlink` works for following one symlink, but I have no honest idea how to do an asynchronous `while` — basically a recursive self-`then`ing Promise — so I cheated and used `fs.readlinkSync` in a `while (true)`.
- In theory, arbitrarily many symlinks can be chained; the way I found to tell when I was done was that `fs.readlinkSync` throws an "EINVAL" error when you give it a non-symlink, so I just `try/catch`’d.
- `npm link` creates an absolute symlink, and `npm link mobx` creates a relative one (🤷🏼‍♂️), so `path.isAbsolute` is handy.
- For relative symlinks, `fs.readlinkSync` appears to return a path relative to the symlink's *parent* folder (e.g. relative to the `node_modules` folder rather than the `node_modules/mobx` folder-cum-symlink), so you need to back up an extra folder when resolving, like: `path.resolve(symlinkLocation, '..', relativeSymlinkTarget);`

**Comment by trotzig** [comment]
Good catch! I am using lerna in https://github.com/Galooshi/happo and some of the code is directly fixing issues I saw early on with symlinked folders/packages. But you're right that symlinked packages with a different name than the original package will cause issues.

**Comment by dcporter** [comment]
Ahhh looks like you're right that Lerna isn't the problem, Lerna is exacerbating the scoped packages problem when they're combined. Let's say there's a monorepo with packages like '@company/package-1', but the packages are in the monorepo at `./packages/package-1`. (The single extra `@company` folder isn't necessary, is a little unfriendly, and is probably not where a developer would think to put them! :) )

**Comment by dcporter** [comment]
This appears to be fixed in #404.
